.github/workflows/ci-test.yml

```diff
@@ -488,9 +491,48 @@ jobs:
         with:
           ref: ${{ github.event.pull_request.head.sha }}
           fetch-depth: 0 #get all the history!!!
-      - uses: ./.github/actions/check-machine
+      - uses: ./.github/actions/build-setup
+      - uses: actions/cache@v2.1.4
+        with:
+          path: "/opt/cargo/git\n/opt/cargo/registry\n/opt/cargo/.package-cache"
+          key: crates-${{ runner.os }}-${{ hashFiles('Cargo.lock') }}
+          restore-keys: "crates-${{ runner.os }}"
+      - name: run unit tests
+        run: |
+          $pre_command && mkdir -p target/junit-reports && cargo nextest --jobs ${max_threads} --unit --failure-output=immediate-final --changed-since "origin/$TARGET_BRANCH" --junit target/junit-reports/unit-test.xml
+          sccache -s
+        env:
+          TARGET_BRANCH: ${{ needs.prepare.outputs.changes-target-branch }}
+      - name: run doctests
+        run: |
+          $pre_command && cargo xtest --doc --jobs ${max_threads} --unit --changed-since "origin/$TARGET_BRANCH"
+          sccache -s
+        env:
+          TARGET_BRANCH: ${{ needs.prepare.outputs.changes-target-branch }}
+      - name: upload unit test results
+        if: always()
+        uses: actions/upload-artifact@v2
         with:
-          webhook-gha-hardware: ${{ secrets.WEBHOOK_GHA_HARDWARE }}
+          name: unit-test-results
+          path: target/junit-reports/unit-test.xml
+      - uses: ./.github/actions/build-teardown
+
+  unit-test-sccache:
```

Comment:
unfortunately there is no way to programmatically select environments for a job, requiring code duplication.

Reply:
do you mind explaining to someone less experienced in this codebase why changing to environment variables for secrets requires these changes with regards to code duplication?

Reply:
gha doesn't let you programmatically choose to use an environment or not in a job, you can programmatically choose to run a job.   hence 2 jobs.  one to use and environment and one not.   the one that uses the environment will have secrets to write to sccache in aws, and the one that doesn't will be able to read from sccache.  the one the writes occurs on push in the "auto" and "canary" branches and performed by bors, the one that doesn't have the secrets is in prs.   normally secrets behaved differently and would not be shared on prs but would be available on any push to any branch, by any user.   the prs would fail if they referenced an environment.